consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftController.java

```diff
@@ -0,0 +1,186 @@
+        }
+        break;
+
+      case IbftV2.PREPARE:
+        final SignedData<PreparePayload> prepareMsg =
+            PrepareMessage.fromMessage(messageData).decode();
+        if (processMessage(prepareMsg, messageData)) {
+          currentHeightManager.handlePrepareMessage(prepareMsg);
+        }
+        break;
+
+      case IbftV2.COMMIT:
+        final SignedData<CommitPayload> commitMsg = CommitMessage.fromMessage(messageData).decode();
+        if (processMessage(commitMsg, messageData)) {
+          currentHeightManager.handleCommitMessage(commitMsg);
+        }
+        break;
+
+      case IbftV2.ROUND_CHANGE:
+        final SignedData<RoundChangePayload> roundChangeMsg =
+            RoundChangeMessage.fromMessage(messageData).decode();
+        if (processRoundMessage(roundChangeMsg)) {
+          currentHeightManager.handleRoundChangeMessage(roundChangeMsg);
+        }
+        break;
+
+      case IbftV2.NEW_ROUND:
+        final SignedData<NewRoundPayload> newRoundMsg =
+            NewRoundMessage.fromMessage(messageData).decode();
+        if (processRoundMessage(newRoundMsg)) {
+          currentHeightManager.handleNewRoundMessage(newRoundMsg);
+        }
+        break;
+
+      default:
+        throw new IllegalArgumentException(
+            "Received message does not conform to any recognised IBFT message structure.");
+    }
+  }
+
+  public void handleNewBlockEvent(final NewChainHead newChainHead) {
+    startNewHeightManager(newChainHead.getNewChainHeadHeader());
+  }
+
+  public void handleBlockTimerExpiry(final BlockTimerExpiry blockTimerExpiry) {
+    currentHeightManager.handleBlockTimerExpiry(blockTimerExpiry.getRoundIndentifier());
+  }
+
+  public void startNewHeightManager(final BlockHeader parentHeader) {
+    currentHeightManager = ibftBlockHeightManagerFactory.create(parentHeader);
+    currentHeightManager.start();
+    final long newChainHeight = currentHeightManager.getChainHeight();
+    futureMessages.getOrDefault(newChainHeight, emptyList()).forEach(this::handleMessage);
+    futureMessages.remove(newChainHeight);
+  }
+
+  public void handleRoundExpiry(final RoundExpiry expired) {
+    currentHeightManager.roundExpired(expired);
+  }
+
+  private boolean processMessage(
+      final SignedData<? extends Payload> msg, final MessageData rawMsg) {
+    final ConsensusRoundIdentifier msgRoundIdentifier = msg.getPayload().getRoundIdentifier();
+    if (isMsgForCurrentHeight(msgRoundIdentifier)) {
+      return isMsgForKnownValidator(msg);
+    } else if (isMsgForFutureChainHeight(msgRoundIdentifier)) {
+      addMessageToFutureMessageBuffer(msgRoundIdentifier.getSequenceNumber(), rawMsg);
+    }
+    return false;
+  }
+
+  private boolean processRoundMessage(final SignedData<? extends Payload> msg) {
+    final ConsensusRoundIdentifier msgRoundIdentifier = msg.getPayload().getRoundIdentifier();
+    return isMsgForCurrentOrFutureChainHeight(msgRoundIdentifier) && isMsgForKnownValidator(msg);
```

Comment:
RoundChange and NewRound Messages are still targetted to current Chain Height, and mus be handled as future messages the same way as Prepare/Commit/Proposal messages - so this function is probably not needed, you can just route RoundChange and NewRound through the standard processMessage function.

Reply:
Yeah your right. Don't know why it did that differently. I'll just change them to use the processMessage method.

Reply:
done